File: archive/JavaScript/Move Zeroes.js
// https://leetcode.com/problems/move-zeroes/

/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 *
 * Given nums = [0, 1, 0, 3, 12], nums should be [1, 3, 12, 0, 0].
 */
var moveZeroes = function(nums) {
  var 
    i = 0, 
    j = 1, 
    len = nums.length
  while (j < len) {
    while (nums[i] != 0 && i < len - 1) {
      i++
    }
    while ((j <= i || nums[j] == 0) && j < len - 1) {
      j++
    }
    [nums[j], nums[i]] = [nums[i], nums[j]]
    j++
  }
  return nums
};
// [1,3,12,0,0]
// [1, 3, 12, 0, 0, 0] => [1, 3, 12, 0, 0]
// [0, 0]

File: 283-move-zeroes/move_zeroes.go
package solution

func moveZeroes(nums []int) {
	j := 0
	for i, n := range nums {
		if n != 0 {
			nums[i], nums[j] = nums[j], nums[i]
			j++
		}
	}
}

func moveZeroes2(nums []int) {
	j := 0
	for _, n := range nums {
		if n != 0 {
			nums[j] = n
			j++
		}
	}
	for j < len(nums) {
		nums[j] = 0
		j++
	}
}
